source/grid/manifold_lib.cc

```diff
@@ -361,80 +361,231 @@ get_tangent_vector (const Point<spacedim> &p1,
-  const int max_iterations = 10;
 
+template <int dim, int spacedim>
+std::pair<double, Tensor<1,spacedim> >
+SphericalManifold<dim,spacedim>::
+guess_new_point(const ArrayView<const Tensor<1,spacedim>> &directions,
+                const ArrayView<const double> &distances,
+                const ArrayView<const double> &weights) const
+{
+  const double tolerance = 1e-10;
   double rho = 0.;
   Tensor<1,spacedim> candidate;
 
-  // Step 1:
   // Perform a simple average ...
-  {
-    double total_weights = 0.;
-    for (unsigned int i = 0; i < n_points; i++)
-      {
-        if (std::abs(1-weights[i])<tolerance)
-          return vertices[i];
-        const Tensor<1, spacedim> direction(vertices[i] - center);
-        rho += direction.norm() * weights[i];
-        candidate += direction * weights[i];
-        total_weights += weights[i];
-      }
-    // ... and normalize if the candidate is different from the origin.
-    const double norm = candidate.norm();
-    if (norm == 0.)
-      return center;
-    candidate /= norm;
-    rho /= total_weights;
-  }
+  double total_weights = 0.;
+  for (unsigned int i = 0; i < directions.size(); i++)
+    {
+      // if one weight is one, return its direction
+      if (std::abs(1-weights[i])<tolerance)
+        return std::make_pair(distances[i],directions[i]);
+
+      rho += distances[i] * weights[i];
+      candidate += directions[i] * weights[i];
+      total_weights += weights[i];
+    }
 
-  // If not in 3D, just use the implementation from PolarManifold
-  // after we verified that the candidate is not the center.
-  if (spacedim<3)
-    return polar_manifold.get_new_point(vertices, weights);
+  // ... and normalize if the candidate is different from the origin.
+  const double norm = candidate.norm();
+  if (norm == 0.)
+    return std::make_pair(0.0,Point<spacedim>());
+  candidate /= norm;
+  rho /= total_weights;
+
+  return std::make_pair(rho,candidate);
+}
+
+
+
+template <int dim, int spacedim>
+Point<spacedim>
+SphericalManifold<dim,spacedim>::
+get_new_point (const ArrayView<const Tensor<1,spacedim>> &directions,
+               const ArrayView<const double> &distances,
+               const ArrayView<const double> &weights,
+               const Point<spacedim> candidate_point) const
+{
+  AssertDimension(directions.size(), distances.size());
+  AssertDimension(directions.size(), weights.size());
+
+  const unsigned int n_merged_points = directions.size();
+  const double tolerance = 1e-10;
+  const int max_iterations = 10;
+
+  // Recover radius and normalized direction from candidate point
+  Tensor <1,spacedim> candidate = candidate_point - center;
+  const double rho = candidate.norm();
+  candidate /= rho;
 
-  // Step 2:
-  // Do Newton-style iterations to improve the estimate.
-  //
   // In this step, we consider all points and directions to be embedded
-  // in a three-dimensional space.
+  // in a three-dimensional space. The case spacedim < 2 was handled in get_new_points()
```

Comment:
So we can specialize for `spacedim==3` and have an `Assert(false, ExcNotImplemented())` for in the general template. This way we can get rid off these annyoing copies.